src/main/scala/org/broadinstitute/dsde/workbench/sam/service/UserService.scala

```diff
@@ -26,20 +25,116 @@ class UserService(val directoryDAO: DirectoryDAO, val cloudExtensions: CloudExte
+      if (user.googleSubjectId.nonEmpty) {
+        directoryDAO.loadUserByGoogleSubjectId(user.googleSubjectId.get, samRequestContext)
+      } else if (user.azureB2CId.nonEmpty) {
+        directoryDAO.loadUserByAzureB2CId(user.azureB2CId.get, samRequestContext)
+      } else {
+        IO(None)
+      }
+    }
+
+  private def assertUserIsNotAlreadyRegistered(user: SamUser, samRequestContext: SamRequestContext): IO[SamUser] =
+    tryToFindUserByCloudId(user, samRequestContext).flatMap {
+      case Some(registeredUser) => IO.raiseError(new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.Conflict, s"user ${registeredUser.email} is already registered")))
+      case None => IO(user)
+    }
+
+  // TODO: Add a simple "updateUser" method to directoryDAO so we can do all this in one call and return the updated user record
+  private def updateUser(existingUserId: WorkbenchUserId, user: SamUser, samRequestContext: SamRequestContext): IO[SamUser] =
+    openTelemetry.time("api.v1.user.updateUser.time", API_TIMING_DURATION_BUCKET) {
+      for {
+        _ <- if (user.googleSubjectId.nonEmpty) directoryDAO.setGoogleSubjectId(existingUserId, user.googleSubjectId.get, samRequestContext) else IO.unit
+        _ <- if (user.azureB2CId.nonEmpty) directoryDAO.setUserAzureB2CId(existingUserId, user.azureB2CId.get, samRequestContext) else IO.unit
+      } yield user.copy(id = existingUserId)
+    }
+
+  private def registerInvitedUser(invitedUser: SamUser, invitedUserId: WorkbenchUserId, samRequestContext: SamRequestContext): IO[SamUser] =
+    openTelemetry.time("api.v1.user.registerInvitedUser.time", API_TIMING_DURATION_BUCKET) {
       for {
-        _ <- validateEmailAddress(user.email, blockedEmailDomains)
-        createdUser <- registerUser(user, samRequestContext)
-        _ <- enableUserInternal(createdUser, samRequestContext)
-        _ <- addToAllUsersGroup(createdUser.id, samRequestContext)
-        userStatus <- getUserStatus(createdUser.id, samRequestContext = samRequestContext)
-        res <- IO
-          .fromOption(userStatus)(new WorkbenchException("getUserStatus returned None after user was created"))
-          .withInfoLogMessage(s"New user ${createdUser.toUserIdInfo} was successfully created")
-      } yield res
+        updatedUser <- updateUser(invitedUserId, invitedUser, samRequestContext)
+        groups <- directoryDAO.listUserDirectMemberships(updatedUser.id, samRequestContext)
+        _ <- IO.fromFuture(IO(cloudExtensions.onGroupUpdate(groups, samRequestContext)))
+      } yield {
+        updatedUser
+      }
     }
 
+  // For now, it looks like createUserInternal does what we need here, but added a new alias method here for naming
+  // consistency and just in case things change more as we are refactoring
+  private def registerBrandNewUser(possibleSamUser: SamUser, samRequestContext: SamRequestContext): IO[SamUser] =
+    createUserInternal(possibleSamUser, samRequestContext)
```

Comment:
It confuses me a little bit to wrap `createUserInternal` for one call, but leave it bare for another call. Would it be clearer instead to rename `createUserInternal` to something more descriptive, like `createUserInDatabase`?